Update managedFields time when field value is modified

glebiller opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The `managedFields.time` is not updated if only the value of a managed field is changed during apply. [According to the doc](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1229-L1233), the time field should be updated during creation, when a field is added or removed, or when a field value is changed.
This PR detects if a managed field has been modified and updates the time accordingly.

#### Which issue(s) this PR fixes:

Fixes #109576

#### Special notes for your reviewer:
I had to remove the unit test validating that when a non-managed field is updated, the time is not changed. It seems impossible to validate that we either have a conflict exception if a non-managed field is updated or the time should be updated.

#### Does this PR introduce a user-facing change?

```release-note
ManagedFields time is correctly updated when the value of a managed field is modified.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
/sig api-machinery
```

glebiller committed:
Update managedFields time when value is modified

glebiller commented:
/assign @apelisse

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater.go:
```diff
@@ -61,15 +64,44 @@ func (f *managedFieldsUpdater) Update(liveObj, newObj runtime.Object, managed Ma
+	}
+
+	compare, _ := liveObjTyped.Compare(newObjTyped)
+	modifiedManagedFields := currentSet.Set().Intersection(compare.Modified)
```
You shouldn't have to perform a full new comparison, especially since we already have one done above (L52):
```golang
object, managed, err := f.fieldManager.Update(liveObj, newObj, managed, self)
```

We've done a similar thing for the `Apply` use-case, as you can see here: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/update.go#L207-L210

We should be able to use the same approach. I think the change would be a lot simpler if you did it that way.

Also, this didn't break any existing test? That's really unfortunate.

apelisse reviewed (commented):
I left a comment, can you also double-check that the behavior with `Apply` is consistent?

Thank you for working on this!

MikeSpreitzer commented:
Thanks for working on this!

glebiller commented:
Following the feedback, I simplified the code and added the following unit tests for both apply & update:
* A change to a managed field should update the time
* No changes to managed fields should not update the time
* A change to a non-managed field should not update the time

### For Update
It looks like [`managed.Fields()[self]`](https://github.com/kubernetes/kubernetes/blob/455bf7c17fe05b553390f830bea129e6bdf5e085/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater.go#L55) is defined only if there is a change to a managed field.
I, therefore, removed the second check to always update the time.

### For Apply
The server-side-apply seems to be functional during my test in a live cluster. With the unit tests, I noticed that the `object` is either nil when the time does not need to be updated or not-nil when the update is required. I removed the extra condition.

## Note for the reviewer
I notice that in the unit test TestNonManagedFieldsUpdateDoesNotModifyTime, after the third Update, the ManagedFields only have the fields from one manager (`fieldmanager_a_test`) and is missing the ManagedFields entry for the second manager ('fieldmanager_b_test'). Is this an intended behavior?

The Apply behavior is different, as the third Apply correctly returns the ManagedFields for both managers with the time changed accordingly.

roycaihw commented:
/triage accepted

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater.go:
```diff
@@ -61,26 +60,22 @@ func (f *managedFieldsUpdater) Update(liveObj, newObj runtime.Object, managed Ma
-			managed.Times()[manager] = &metav1.Time{Time: time.Now().UTC()}
-		}
+
+		managed.Times()[manager] = &metav1.Time{Time: time.Now().UTC()}
```
That's actually more consistent with the comment line 54 above, thanks

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -29,42 +30,255 @@ import (
+	newManagedFields := f.ManagedFields()
+
+	if reflect.DeepEqual(previousManagedFields, newManagedFields) {
+		t.Errorf("ManagedFields time has not been updated:\n%v", newManagedFields)
```
That's not really what you're comparing so that error might be misleading (also applies below)

glebiller replied:
I am not sure to understand it, should it directly compare the Time field of both entries?
Current, if the entries are identical, that means the time has not been updated.

glebiller replied:
I added a test case named `TestTakingOverManagedFieldsDuringUpdateDoesNotModifyPreviousManagerTime` for the Update.

apelisse replied:
You're comparing `ManagedFields`, but the error reads `time has not been updated`, which is slightly misleading.

apelisse replied:
But the code looks good now, thanks

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -78,6 +292,26 @@ func (NoopManager) Update(liveObj, newObj runtime.Object, managed Managed, manag
 	return nil, nil, nil
 }
 
+func updateObject(t *testing.T, f *TestFieldManager, fieldManagerName string, object []byte) {
```
Can you leave `t` out of these two functions (update and apply), return an error from these?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -29,42 +30,255 @@ import (
+		newEntries[entry.Manager] = entry
+	}
+
+	if reflect.DeepEqual(previousManagedFields, newManagedFields) {
```
This test isn't really needed since it'll fail below anyway, right? (same thing applies to the next test for apply)

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -29,42 +30,255 @@ import (
+		t.Errorf("FieldManager A ManagedFields time has not been updated:\n%v", newEntries["fieldmanager_a_test"])
+	}
+	if _, ok := newEntries["fieldmanager_b_test"]; ok {
+		t.Errorf("FieldManager B ManagedFields has changed changed:\n%v", newEntries["fieldmanager_b_test"])
```
nit, 
```suggestion
		t.Errorf("FieldManager B ManagedFields has changed:\n%v", newEntries["fieldmanager_b_test"])
```

apelisse reviewed (commented):
One thing that we have in the doc but isn't tested here, is that when a field is removed from a set (because it's stolen by someone else), that set's timestamp doesn't change, i.e.

In mostly pseudo-code:
```golang
applyObject("managerA", `{fieldA: true, fieldB: true}`)
time.Sleep(1)
applyObject("managerB", `{fieldA: false}`)
// Check that managerA timestamp hasn't changed.
```

glebiller commented:
> One thing that we have in the doc but isn't tested here, is that when a field is removed from a set (because it's stolen by someone else), that set's timestamp doesn't change, i.e.
> 
> In mostly pseudo-code:
> 
> ```go
> applyObject("managerA", `{fieldA: true, fieldB: true}`)
> time.Sleep(1)
> applyObject("managerB", `{fieldA: false}`)
> // Check that managerA timestamp hasn't changed.
> ```

According to the [Server-Side-Apply doc](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts), this is not possible because it raises a conflict error. I tried creating a unit test for that case, but I got the following error:
`failed to apply object: Apply failed with 1 conflict: conflict with "fieldmanager_a_test": .data.key_b`

The use case that could be tested, is during the Update instead of Apply.

apelisse commented:
> According to the [Server-Side-Apply doc](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts), this is not possible because it raises a conflict error. I tried creating a unit test for that case, but I got the following error: `failed to apply object: Apply failed with 1 conflict: conflict with "fieldmanager_a_test": .data.key_b`
> 
> The use case that could be tested, is during the Update instead of Apply.

There is a "Force" field to steal ownership with `Apply`, which will let you test the same scenario!

apelisse commented:
Thanks, LGTM once you add the `TestTakingOverManagedFieldsDuringApplyDoesNotModifyPreviousManagerTime`

glebiller commented:
That's on me, I should have though about the force as it's similar to the kubectl command line :)
I added the unit test and all the UT are now passing! Thanks for your help @apelisse

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -78,6 +433,28 @@ func (NoopManager) Update(liveObj, newObj runtime.Object, managed Managed, manag
+	return nil
+}
+
+func applyObject(f *TestFieldManager, fieldManagerName string, object []byte) error {
```
I would maybe have taken the force parameter in that method, but that's probably fine for now.

apelisse reviewed (commented):
/lgtm
/approve

Thanks!
